certora/specs/AccessManager.spec

```diff
@@ -0,0 +1,766 @@
+    selector == to_bytes4(sig:updateAuthority(address,address).selector              ) ||
+    selector == to_bytes4(sig:setTargetClosed(address,bool).selector                 ) ||
+    selector == to_bytes4(sig:setTargetFunctionRole(address,bytes4[],uint64).selector) ||
+    selector == to_bytes4(sig:grantRole(uint64,address,uint32).selector              ) ||
+    selector == to_bytes4(sig:revokeRole(uint64,address).selector                    );
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: executionId must be clean when not in the middle of a call                                               │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant cleanExecutionId()
+    executionId() == to_bytes32(0);
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: public role                                                                                              │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant publicRole(env e, address account)
+    hasRole_1(e, PUBLIC_ROLE(), account)        &&
+    hasRole_2(e, PUBLIC_ROLE(), account)   == 0 &&
+    getAccess_1(e, PUBLIC_ROLE(), account) == 0 &&
+    getAccess_2(e, PUBLIC_ROLE(), account) == 0 &&
+    getAccess_3(e, PUBLIC_ROLE(), account) == 0 &&
+    getAccess_4(e, PUBLIC_ROLE(), account) == 0;
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: hasRole is consistent with getAccess                                                                     │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant hasRoleGetAccessConsistency(env e, uint64 roleId, address account)
+    hasRole_1(e, roleId, account) == (roleId == PUBLIC_ROLE() || isSetAndPast(e, getAccess_1(e, roleId, account))) &&
+    hasRole_2(e, roleId, account) == getAccess_2(e, roleId, account);
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Functions: canCall, canCallExtended, getAccess, hasRole, isTargetClosed and getTargetFunctionRole do NOT revert     │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule noRevert(env e) {
+    require nonpayable(e);
+    require envsanity(e);
+
+    address caller;
+    address target;
+    bytes   data;
+    bytes4  selector;
+    uint64  roleId;
+
+    canCall@withrevert(e, caller, target, selector);
+    assert !lastReverted;
+
+    canCallExtended@withrevert(e, caller, target, data);
+    assert !lastReverted;
```

Comment:
This does fail. I'm not sure why.

Reply:
![Capture d’écran du 2023-09-22 17-49-57](https://github.com/OpenZeppelin/openzeppelin-contracts/assets/2432299/c88147a3-ff15-41bb-a2f2-570b671b06b2)